[API] Writable start and due date

NobodysNightmare opened the pull request:
## OpenProject work packages
- https://community.openproject.org/work_packages/17417
- https://community.openproject.org/work_packages/17418

NobodysNightmare committed:
represent start and due date as date only

also add tests for output formats of dates and times in work package

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://feature-writeable-start-and-end.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-writeable-start-and-end.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
allow PATCHing start and due date

- adding new errors
- adding more tests
- extending the specification for error conditions

NobodysNightmare committed:
fix integrative tests after PATCH changes

NobodysNightmare committed:
include dates in schema

NobodysNightmare committed:
styling

NobodysNightmare committed:
use shared example for payload spec

NobodysNightmare committed:
add missing integration tests

ulferts commented on lib/api/v3/work_packages/form/work_package_payload_representer.rb:
```diff
@@ -108,6 +121,27 @@ def work_package_attribute_links_representer(represented)
+            end
+
+            date_only
+          end
```
@NobodysNightmare I'd like to get the payload representer as small as possible. For that I propose moving the date parsing to some helper object that all representers (or whoever) can reuse. Otherwise I see the danger of us having to reimplement this time and time again in the various payload representers and then having to switch the format on the date string.

Because of said format, it should ideally also include the getter part. That would also take care of the repetition of
`<date attribute>.to_date.iso8601 unless <date attribute>.nil?`
in four places in this PR alone.

And if that is not enough for you to buy it, it will also cut down the required specs.

ulferts commented on config/locales/en.yml:
```diff
@@ -1665,6 +1665,7 @@ en:
       multiple_errors: "Multiple fields violated their constraints."
       parse_error: "The request body was neither empty, nor did it contain a single JSON object."
       writing_read_only_attributes: "You must not write a read-only attribute."
+      invalid_format: "Invalid format for property '%{property}': Expected format like '%{expected_format}', but got '%{actual}'."
```
I was a bit startled when I accidentally inserted "2014-02-30" to test this because the format looks valid. I could also have sent "9999-99-99". Do you think it makes sense to also state what "Type" is required here? Then a user could look up the valid values in the API documentation as soon as the types are described there.

NobodysNightmare replied:
Even worse: ISO 8601 allows more than this specific format... However, I thought it was better than writing:

> Expected format like 'ISO 8601'

I _wanted_ to make the message generic enough to be reusable for other formating errors...

So better something like:

> Invalid format for property 'startDate': Expected input format of 'Date', but could not parse '2014-02-30' as such.

Would that be better?

ulferts commented:
Aside from my two remarks, this is looking good to merge.

NobodysNightmare committed:
refactor output of dates and times

- all (?) APIv3 resources now use a common formatter
- all APIv3 resources now use common names for "createdAt" and "updatedAt"
- existing tests were adapted when they were wrong before (according to specs)
- no unit tests for common formatter extracted yet

NobodysNightmare committed:
unit test DateTimeFormatter

- users of the formatter are only tested to call the formatter

NobodysNightmare committed:
use shared examples in more tests

NobodysNightmare committed:
satisfy hound

also make the private modifier less useless

NobodysNightmare commented:
@ulferts Only a decision on good error messages left I guess...

NobodysNightmare committed:
fix parameter name in test

NobodysNightmare committed:
use Time instead of DateTime in factories

- this seems to be the mapping rails uses
- this matters because ActiveSupport fails to return a Time from "to_time"

NobodysNightmare commented:
That looks like two false negative tests to me :(

NobodysNightmare committed:
unify error messages

NobodysNightmare committed:
fix integrative tests after rewording of errors

ulferts commented on spec/lib/api/v3/work_packages/work_package_representer_spec.rb:
```diff
@@ -35,12 +35,15 @@
     FactoryGirl.create(:user,  member_in_project: project, member_through_role: role)
   end
   let(:current_user) { member }
+  let(:current_user) { member }
```
Why? Tell me, tell me, tell me...

NobodysNightmare replied:
I prepared this as a [duck](http://blog.codinghorror.com/new-programming-jargon/) for you... Congratulations. You found it!

ulferts commented on lib/api/v3/utilities/date_time_formatter.rb:
```diff
@@ -0,0 +1,94 @@
+          Duration.new(hours_and_minutes(hours)).iso8601
+        end
+
+        private_class_method
```
This does not do what you expect

```
pry(main)> API::V3::Utilities::DateTimeFormatter::hours_and_minutes(56.54)
=> {:hours=>56, :minutes=>32.39999999999995}
```

What you mean to do is:

```
private_class_method :hours_and_minutes
```

I haven't been able to find good source on why this behaves different than the `private` method.

While you are at it: I think that the parameter name `hours` in the method is wrong as it is later split into hours and minutes.

NobodysNightmare replied:
I don't think that `hours` is too wrong, because both statements are equally valid:
- we meet in 1.5 hours
- we meet in one hour and thirty minutes

At first I was tempted to write a name that indicates hours can be float (`fractional_hours`, anyone?). But I guess it should not matter if you give me a float or an int, should it?

ulferts commented:
@NobodysNightmare just reminder of what we talked about: We could decrease the dependency if you where to include the DateTimeFormatter in API::Decorators::Single and there wrap it in a method.

@0xF013 in the PR @NobodysNightmare changed the estimatedTime property so that it returns null if the value is not set. This currently breaks the front end. So if estimatedTime is null, it should be counted as not set ("-" displayed) if it is 0 it should be counted as such ("0" displayed).

0xF013 commented:
@ulferts should I take over this branch?

ulferts commented:
@0xF013 yes, please add to it. I think the changes @NobodysNightmare will make are minor and should not impede the front end work.

NobodysNightmare committed:
remove errorneous double line

NobodysNightmare committed:
make that method really really private

NobodysNightmare commented:
@0xF013 You can have it... either wait until my next batch of commits arrives or just start. In the latter case, the one of us that takes longer to develop will have to rebase. I think we should not be conflicting eachother.

edit: I did not see the comment of @ulferts until sending my comment -.- Whatever, I jsut said the same ^^

0xF013 committed:
treat null as empty
